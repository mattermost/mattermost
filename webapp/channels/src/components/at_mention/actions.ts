// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {UserProfile} from '@mattermost/types/users';

import {getMissingProfilesByUsernames} from 'mattermost-redux/actions/users';

import {getPotentialMentionsForName} from 'utils/post_utils';

import type {ActionFuncAsync} from 'types/store';

export function getMissingMentionedUsers(text: string): ActionFuncAsync<Array<UserProfile['username']>> {
    return getMissingProfilesByUsernames(getPotentialMentionsForName(text));
}
