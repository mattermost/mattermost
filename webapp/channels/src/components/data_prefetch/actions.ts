// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {Channel, ChannelMembership} from '@mattermost/types/channels';
import type {RelationOneToOne} from '@mattermost/types/utilities';

import {isChannelMuted} from 'mattermost-redux/utils/channel_utils';
import {memoizeResult} from 'mattermost-redux/utils/helpers';

enum Priority {
    high = 1,
    medium,
    low
}

enum PrefetchLimits {
    mentionMax = 10,
    unreadMax = 20,
}

// function to return a queue obj with priotiy as key and array of channelIds as values.
// high priority has channels with mentions
// medium priority has channels with unreads
// <10 unread channels. Prefetch everything.
// 10-20 unread. Prefetch only mentions, capped to 10.
// >20 unread. Don't prefetch anything.
export const prefetchQueue = memoizeResult((
    unreadChannels: Channel[],
    memberships: RelationOneToOne<Channel, ChannelMembership>,
    collapsedThreads: boolean,
) => {
    const unreadChannelsCount = unreadChannels.length;
    let result: {
        1: string[];
        2: string[];
        3: string[];
    } = {
        [Priority.high]: [], // 1 being high priority requests
        [Priority.medium]: [],
        [Priority.low]: [], //TODO: add chanenls such as fav.
    };
    if (!unreadChannelsCount || unreadChannelsCount > PrefetchLimits.unreadMax) {
        return result;
    }
    for (const channel of unreadChannels) {
        const channelId = channel.id;
        const membership = memberships[channelId];

        if (unreadChannelsCount >= PrefetchLimits.mentionMax && result[Priority.high].length >= PrefetchLimits.mentionMax) {
            break;
        }

        // TODO We check for muted channels 3 times here: getUnreadChannels checks it, this checks it, and the mark_unread
        // check below is equivalent to checking if its muted.
        if (membership && !isChannelMuted(membership)) {
            if (collapsedThreads ? membership.mention_count_root : membership.mention_count) {
                result = {
                    ...result,
                    [Priority.high]: [...result[Priority.high], channelId],
                };
            } else if (
                membership.notify_props &&
                membership.notify_props.mark_unread !== 'mention' &&
                unreadChannelsCount < PrefetchLimits.mentionMax
            ) {
                result = {
                    ...result,
                    [Priority.medium]: [...result[Priority.medium], channelId],
                };
            }
        }
    }
    return result;
});
