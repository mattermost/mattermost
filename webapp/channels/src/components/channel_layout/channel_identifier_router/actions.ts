// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {History} from 'history';

import type {Channel} from '@mattermost/types/channels';
import type {GlobalState} from '@mattermost/types/store';

import {joinChannel, getChannelByNameAndTeamName, getChannelMember, markGroupChannelOpen, fetchChannelsAndMembers} from 'mattermost-redux/actions/channels';
import {getUser, getUserByUsername, getUserByEmail} from 'mattermost-redux/actions/users';
import {Client4} from 'mattermost-redux/client';
import {getChannelByName, getOtherChannels, getChannel, getChannelsNameMapInTeam, getRedirectChannelNameForTeam} from 'mattermost-redux/selectors/entities/channels';
import {getTeamByName} from 'mattermost-redux/selectors/entities/teams';
import {getCurrentUser, getCurrentUserId, getUserByUsername as selectUserByUsername, getUser as selectUser, getUserByEmail as selectUserByEmail} from 'mattermost-redux/selectors/entities/users';
import * as UserUtils from 'mattermost-redux/utils/user_utils';

import {openDirectChannelToUserId} from 'actions/channel_actions';
import * as GlobalActions from 'actions/global_actions';

import {joinPrivateChannelPrompt} from 'utils/channel_utils';
import {Constants} from 'utils/constants';
import * as Utils from 'utils/utils';

import type {ActionFuncAsync} from 'types/store';

import type {Match, MatchAndHistory} from './channel_identifier_router';

const LENGTH_OF_ID = 26;
const LENGTH_OF_GROUP_ID = 40;
const LENGTH_OF_USER_ID_PAIR = 54;
const USER_ID_PAIR_REGEXP = new RegExp(`^[a-zA-Z0-9]{${LENGTH_OF_ID}}__[a-zA-Z0-9]{${LENGTH_OF_ID}}$`);

export function onChannelByIdentifierEnter({match, history}: MatchAndHistory): ActionFuncAsync {
    return async (dispatch, getState) => {
        const state = getState();
        const {path, identifier, team} = match.params;

        if (!identifier) {
            return {data: undefined};
        }

        const teamObj = getTeamByName(state, team);
        if (!teamObj) {
            return {data: undefined};
        }

        const channelPath = await getPathFromIdentifier(state, path, identifier);

        switch (channelPath) {
        case 'channel_name':
            dispatch(goToChannelByChannelName(match, history));
            break;
        case 'channel_id':
            dispatch(goToChannelByChannelId(match, history));
            break;
        case 'group_channel_group_id':
            dispatch(goToGroupChannelByGroupId(match, history));
            break;
        case 'direct_channel_username':
            dispatch(goToDirectChannelByUsername(match, history));
            break;
        case 'direct_channel_email':
            dispatch(goToDirectChannelByEmail(match, history));
            break;
        case 'direct_channel_user_ids':
            dispatch(goToDirectChannelByUserIds(match, history));
            break;
        case 'direct_channel_user_id':
            dispatch(goToDirectChannelByUserId(match, history, identifier));
            break;
        case 'error':
            await dispatch(fetchChannelsAndMembers(teamObj!.id));
            handleError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
            break;
        }
        return {data: undefined};
    };
}

export async function getPathFromIdentifier(state: GlobalState, path: string, identifier: string) {
    if (path === 'channels') {
        // It's hard to tell an ID apart from a channel name of the same length, so check first if
        // the identifier matches a channel that we have
        const channelsByName = getChannelByName(state, identifier);
        const moreChannelsByName = getOtherChannels(state).find((chan) => chan.name === identifier);

        if (identifier.length === LENGTH_OF_ID) {
            if (!channelsByName && !moreChannelsByName) {
                try {
                    await Client4.getChannel(identifier);
                    return 'channel_id';
                } catch (e) {
                    if (e.status_code === 404) {
                        return 'channel_name';
                    }
                    return 'error';
                }
            }
            return 'channel_name';
        } else if (
            (!channelsByName && !moreChannelsByName && identifier.length === LENGTH_OF_GROUP_ID) ||
            (
                (channelsByName && channelsByName.type === Constants.GM_CHANNEL) ||
                (moreChannelsByName && moreChannelsByName.type === Constants.GM_CHANNEL)
            )
        ) {
            return 'group_channel_group_id';
        } else if (isDirectChannelIdentifier(identifier)) {
            return 'direct_channel_user_ids';
        }
        return 'channel_name';
    } else if (path === 'messages') {
        if (identifier.indexOf('@') === 0) {
            return 'direct_channel_username';
        } else if (identifier.indexOf('@') > 0) {
            return 'direct_channel_email';
        } else if (identifier.length === LENGTH_OF_ID) {
            return 'direct_channel_user_id';
        } else if (identifier.length === LENGTH_OF_GROUP_ID) {
            return 'group_channel_group_id';
        }
        return 'error';
    }

    return 'error';
}

export function goToChannelByChannelId(match: Match, history: History): ActionFuncAsync {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const channelId = identifier.toLowerCase();

        let channel = getChannel(state, channelId);
        const member = state.entities.channels.myMembers[channelId];
        const teamObj = getTeamByName(state, team);
        if (!channel || !member) {
            const dispatchResult = await dispatch(joinChannel(getCurrentUserId(state), teamObj!.id, channelId, ''));
            if ('error' in dispatchResult) {
                await dispatch(fetchChannelsAndMembers(teamObj!.id));
                handleChannelJoinError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
                return {data: undefined};
            }
            channel = dispatchResult.data!.channel;
        }

        if (channel.type === Constants.DM_CHANNEL) {
            dispatch(goToDirectChannelByUserId(match, history, Utils.getUserIdFromChannelId(channel.name, getCurrentUserId(state))));
        } else if (channel.type === Constants.GM_CHANNEL) {
            history.replace(`/${team}/messages/${channel.name}`);
        } else {
            history.replace(`/${team}/channels/${channel.name}`);
        }
        return {data: undefined};
    };
}

export function goToChannelByChannelName(match: Match, history: History): ActionFuncAsync {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const channelName = identifier.toLowerCase();

        const teamObj = getTeamByName(state, team);
        if (!teamObj) {
            return {data: undefined};
        }

        let channel = getChannelsNameMapInTeam(state, teamObj!.id)[channelName];
        if (!channel) {
            const getChannelDispatchResult = await dispatch(getChannelByNameAndTeamName(team, channelName, true));
            if ('data' in getChannelDispatchResult) {
                channel = getChannelDispatchResult.data!;
            }
        }

        let member;
        if (channel) {
            member = state.entities.channels.myMembers[channel.id];
            if (!member) {
                const membership = await dispatch(getChannelMember(channel.id, getCurrentUserId(state)));
                if ('data' in membership) {
                    member = membership.data;
                }
            }
        }

        if (!channel || !member) {
            // Prompt system admin before joining the private channel
            const user = getCurrentUser(getState());
            const isSystemAdmin = UserUtils.isSystemAdmin(user?.roles);
            if (isSystemAdmin) {
                if (channel?.type === Constants.PRIVATE_CHANNEL) {
                    const joinPromptResult = await dispatch(joinPrivateChannelPrompt(teamObj, channel.display_name));
                    if ('data' in joinPromptResult && !joinPromptResult.data!.join) {
                        return {data: undefined};
                    }
                }
            }

            const joinChannelDispatchResult = await dispatch(joinChannel(getCurrentUserId(state), teamObj!.id, '', channelName));
            if ('error' in joinChannelDispatchResult) {
                if (!channel) {
                    const getChannelDispatchResult = await dispatch(getChannelByNameAndTeamName(team, channelName, true));
                    if ('error' in getChannelDispatchResult || getChannelDispatchResult.data!.delete_at === 0) {
                        await dispatch(fetchChannelsAndMembers(teamObj!.id));
                        handleChannelJoinError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
                        return {data: undefined};
                    }
                    channel = getChannelDispatchResult.data!;
                }
            } else {
                channel = joinChannelDispatchResult.data!.channel;
            }
        }

        if (channel.type === Constants.DM_CHANNEL) {
            dispatch(goToDirectChannelByUserIds(match, history));
        } else if (channel.type === Constants.GM_CHANNEL) {
            history.replace(`/${team}/messages/${channel.name}`);
        } else {
            doChannelChange(channel);
        }
        return {data: undefined};
    };
}

function goToDirectChannelByUsername(match: Match, history: History): ActionFuncAsync {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const username = identifier.slice(1, identifier.length).toLowerCase();
        const teamObj = getTeamByName(state, team);

        let user = selectUserByUsername(state, username);
        if (!user) {
            const dispatchResult = await dispatch(getUserByUsername(username));
            if ('error' in dispatchResult) {
                await dispatch(fetchChannelsAndMembers(teamObj!.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
                return {data: undefined};
            }
            user = dispatchResult.data!;
        }

        const directChannelDispatchRes = await dispatch(openDirectChannelToUserId(user.id));
        if ('error' in directChannelDispatchRes) {
            await dispatch(fetchChannelsAndMembers(teamObj!.id));
            handleError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
            return {data: undefined};
        }

        doChannelChange(directChannelDispatchRes.data!);
        return {data: undefined};
    };
}

export function goToDirectChannelByUserId(match: Match, history: History, userId: string): ActionFuncAsync {
    return async (dispatch, getState) => {
        const state = getState();
        const {team} = match.params;
        const teamObj = getTeamByName(state, team);

        let user = selectUser(state, userId);
        if (!user) {
            const dispatchResult = await dispatch(getUser(userId));
            if ('error' in dispatchResult) {
                await dispatch(fetchChannelsAndMembers(teamObj!.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
                return {data: undefined};
            }
            user = dispatchResult.data!;
        }

        history.replace(`/${team}/messages/@${user.username}`);
        return {data: undefined};
    };
}

export function goToDirectChannelByUserIds(match: Match, history: History): ActionFuncAsync {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const userId = Utils.getUserIdFromChannelId(identifier.toLowerCase(), getCurrentUserId(getState()));
        const teamObj = getTeamByName(state, team);

        let user = selectUser(state, userId);
        if (!user) {
            const dispatchResult = await dispatch(getUser(userId));
            if ('error' in dispatchResult) {
                await dispatch(fetchChannelsAndMembers(teamObj!.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
                return {data: undefined};
            }
            user = dispatchResult.data!;
        }

        history.replace(`/${team}/messages/@${user.username}`);
        return {data: undefined};
    };
}

export function goToDirectChannelByEmail(match: Match, history: History): ActionFuncAsync {
    return async (dispatch, getState) => {
        const state = getState();
        const {team, identifier} = match.params;
        const email = identifier.toLowerCase();
        const teamObj = getTeamByName(state, team);

        let user = selectUserByEmail(state, email);
        if (!user) {
            const dispatchResult = await dispatch(getUserByEmail(email));
            if ('error' in dispatchResult) {
                await dispatch(fetchChannelsAndMembers(teamObj!.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
                return {data: undefined};
            }
            user = dispatchResult.data!;
        }

        history.replace(`/${team}/messages/@${user.username}`);
        return {data: undefined};
    };
}

function goToGroupChannelByGroupId(match: Match, history: History): ActionFuncAsync {
    return async (dispatch, getState) => {
        const state = getState();
        const {identifier, team} = match.params;
        const groupId = identifier.toLowerCase();

        history.replace(match.url.replace('/channels/', '/messages/'));

        let channel = getChannelByName(state, groupId);
        const teamObj = getTeamByName(state, team);
        if (!channel) {
            const dispatchResult = await dispatch(joinChannel(getCurrentUserId(state), teamObj!.id, '', groupId));
            if ('error' in dispatchResult) {
                await dispatch(fetchChannelsAndMembers(teamObj!.id));
                handleError(match, history, getRedirectChannelNameForTeam(state, teamObj!.id));
                return {data: undefined};
            }
            channel = dispatchResult.data!.channel;
        }

        dispatch(markGroupChannelOpen(channel!.id));

        doChannelChange(channel!);
        return {data: undefined};
    };
}

function doChannelChange(channel: Channel) {
    GlobalActions.emitChannelClickEvent(channel);
}

function handleError(match: Match, history: History, defaultChannel: string) {
    const {team} = match.params;
    history.push(team ? `/${team}/channels/${defaultChannel}` : '/');
}

function isDirectChannelIdentifier(identifier: string) {
    return identifier.length === LENGTH_OF_USER_ID_PAIR && USER_ID_PAIR_REGEXP.test(identifier);
}

async function handleChannelJoinError(match: Match, history: History, defaultChannel: string) {
    const {team} = match.params;
    history.push(team ? `/error?type=channel_not_found&returnTo=/${team}/channels/${defaultChannel}` : '/');
}
