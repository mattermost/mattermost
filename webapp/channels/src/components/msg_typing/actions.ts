// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {GlobalState} from '@mattermost/types/store';

import {getMissingProfilesByIds, getStatusesByIds} from 'mattermost-redux/actions/users';
import {General, Preferences, WebsocketEvents} from 'mattermost-redux/constants';
import {getIsUserStatusesConfigEnabled} from 'mattermost-redux/selectors/entities/common';
import {getConfig, isPerformanceDebuggingEnabled} from 'mattermost-redux/selectors/entities/general';
import {getBool} from 'mattermost-redux/selectors/entities/preferences';
import {getCurrentUserId, getStatusForUserId} from 'mattermost-redux/selectors/entities/users';

import type {ActionFuncAsync, ThunkActionFunc} from 'types/store';

function getTimeBetweenTypingEvents(state: GlobalState) {
    const config = getConfig(state);

    return config.TimeBetweenUserTypingUpdatesMilliseconds === undefined ? 0 : parseInt(config.TimeBetweenUserTypingUpdatesMilliseconds, 10);
}

export function userStartedTyping(userId: string, channelId: string, rootId: string, now: number): ThunkActionFunc<void> {
    return (dispatch, getState) => {
        const state = getState();

        if (
            isPerformanceDebuggingEnabled(state) &&
            getBool(state, Preferences.CATEGORY_PERFORMANCE_DEBUGGING, Preferences.NAME_DISABLE_TYPING_MESSAGES)
        ) {
            return;
        }

        dispatch({
            type: WebsocketEvents.TYPING,
            data: {
                id: channelId + rootId,
                userId,
                now,
            },
        });

        // Ideally this followup loading would be done by someone else
        dispatch(fillInMissingInfo(userId));

        setTimeout(() => {
            dispatch(userStoppedTyping(userId, channelId, rootId, now));
        }, getTimeBetweenTypingEvents(state));
    };
}

function fillInMissingInfo(userId: string): ActionFuncAsync {
    return async (dispatch, getState) => {
        const state = getState();
        const currentUserId = getCurrentUserId(state);
        const enabledUserStatuses = getIsUserStatusesConfigEnabled(state);

        if (userId !== currentUserId) {
            const result = await dispatch(getMissingProfilesByIds([userId]));
            if (result.data && result.data.length > 0) {
                // Already loaded the user status
                return {data: false};
            }
        }

        const status = getStatusForUserId(state, userId);
        if (status !== General.ONLINE && enabledUserStatuses) {
            dispatch(getStatusesByIds([userId]));
        }

        return {data: true};
    };
}

export function userStoppedTyping(userId: string, channelId: string, rootId: string, now: number) {
    return {
        type: WebsocketEvents.STOP_TYPING,
        data: {
            id: channelId + rootId,
            userId,
            now,
        },
    };
}
