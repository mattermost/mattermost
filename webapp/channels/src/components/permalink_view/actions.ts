// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import type {Channel} from '@mattermost/types/channels';
import type {Post} from '@mattermost/types/posts';

import {getChannel, getChannelMember, selectChannel, joinChannel, getChannelStats} from 'mattermost-redux/actions/channels';
import {getPostThread} from 'mattermost-redux/actions/posts';
import {getMissingProfilesByIds} from 'mattermost-redux/actions/users';
import {Client4} from 'mattermost-redux/client';
import {getCurrentChannel, getChannel as getChannelFromRedux} from 'mattermost-redux/selectors/entities/channels';
import {isCollapsedThreadsEnabled} from 'mattermost-redux/selectors/entities/preferences';
import {getCurrentTeam, getTeam} from 'mattermost-redux/selectors/entities/teams';
import {getCurrentUser} from 'mattermost-redux/selectors/entities/users';
import {getUserIdFromChannelName} from 'mattermost-redux/utils/channel_utils';
import {isSystemAdmin} from 'mattermost-redux/utils/user_utils';

import {loadChannelsForCurrentUser} from 'actions/channel_actions';
import {loadNewDMIfNeeded, loadNewGMIfNeeded} from 'actions/user_actions';
import {selectPostAndHighlight} from 'actions/views/rhs';

import {getHistory} from 'utils/browser_history';
import {joinPrivateChannelPrompt} from 'utils/channel_utils';
import {ActionTypes, Constants, ErrorPageTypes} from 'utils/constants';
import {isComment, getPostURL} from 'utils/post_utils';

import type {ActionFuncAsync, ThunkActionFunc} from 'types/store';

let privateChannelJoinPromptVisible = false;

type Option = {
    skipRedirectReplyPermalink: boolean;
}

function focusRootPost(post: Post, channel: Channel): ActionFuncAsync {
    return async (dispatch, getState) => {
        const postURL = getPostURL(getState(), post);

        dispatch(selectChannel(channel.id));
        dispatch({
            type: ActionTypes.RECEIVED_FOCUSED_POST,
            data: post.id,
            channelId: channel.id,
        });

        getHistory().replace(postURL);
        return {data: true};
    };
}

function focusReplyPost(post: Post, channel: Channel, teamId: string, returnTo: string, option: Option): ActionFuncAsync {
    return async (dispatch, getState) => {
        const {data} = await dispatch(getPostThread(post.root_id));

        if (data!.first_inaccessible_post_time) {
            getHistory().replace(`/error?type=${ErrorPageTypes.CLOUD_ARCHIVED}&returnTo=${returnTo}`);
            return {data: false};
        }

        const state = getState();

        const team = getTeam(state, channel.team_id || teamId);
        const currentChannel = getCurrentChannel(state);

        const sameTeam = currentChannel && currentChannel.team_id === team?.id;

        const {skipRedirectReplyPermalink} = option;

        if (!sameTeam) {
            dispatch(selectChannel(channel.id));
        }

        if (sameTeam && returnTo && !skipRedirectReplyPermalink) {
            getHistory().replace(returnTo);
        } else if (!sameTeam || !skipRedirectReplyPermalink) {
            const postURL = getPostURL(state, post);
            getHistory().replace(postURL);
        }

        dispatch(selectPostAndHighlight(post));
        return {data: true};
    };
}

export function focusPost(postId: string, returnTo = '', currentUserId: string, option: Option = {skipRedirectReplyPermalink: false}): ThunkActionFunc<Promise<void>> {
    return async (dispatch, getState) => {
        // Ignore if prompt is still visible
        if (privateChannelJoinPromptVisible) {
            return;
        }

        let postInfo;
        try {
            postInfo = await Client4.getPostInfo(postId);
        } catch (e) {
            getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
            return;
        }

        const state = getState();
        const currentTeam = getCurrentTeam(state);
        if (!currentTeam) {
            return;
        }

        if (!postInfo.has_joined_channel) {
            // Prompt system admin before joining the private channel
            const user = getCurrentUser(state);
            if (postInfo.channel_type === Constants.PRIVATE_CHANNEL && isSystemAdmin(user.roles)) {
                privateChannelJoinPromptVisible = true;
                const joinPromptResult = await dispatch(joinPrivateChannelPrompt(currentTeam, postInfo.channel_display_name));
                privateChannelJoinPromptVisible = false;
                if ('data' in joinPromptResult && !joinPromptResult.data!.join) {
                    return;
                }
            }
            await dispatch(joinChannel(currentUserId, '', postInfo.channel_id));
        }

        const {data: threadData} = await dispatch(getPostThread(postId));

        if (!threadData) {
            getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
            return;
        }

        if (threadData.first_inaccessible_post_time) {
            getHistory().replace(`/error?type=${ErrorPageTypes.CLOUD_ARCHIVED}&returnTo=${returnTo}`);
            return;
        }

        const isCollapsed = isCollapsedThreadsEnabled(state);

        const channelId = threadData.posts[threadData.order[0]].channel_id;
        let channel = getChannelFromRedux(state, channelId);

        if (!channel) {
            const {data: channelData} = await dispatch(getChannel(channelId));

            if (!channelData) {
                getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
                return;
            }

            channel = channelData;
        }

        const teamId = channel.team_id || currentTeam.id;
        let myMember = state.entities.channels.myMembers[channelId];

        if (!myMember) {
            // If it's a DM or GM channel and we don't have a channel member for it already, user is not a member
            if (channel.type === Constants.DM_CHANNEL || channel.type === Constants.GM_CHANNEL) {
                getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
                return;
            }

            const membership = await dispatch(getChannelMember(channel.id, currentUserId));
            if ('data' in membership) {
                myMember = membership.data!;
            }

            if (!myMember) {
                getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
                return;
            }
        }

        if (channel.team_id && channel.team_id !== teamId) {
            getHistory().replace(`/error?type=${ErrorPageTypes.PERMALINK_NOT_FOUND}&returnTo=${returnTo}`);
            return;
        }

        if (channel && channel.type === Constants.DM_CHANNEL) {
            const userId = getUserIdFromChannelName(currentUserId, channel.name);
            await dispatch(getMissingProfilesByIds([userId]));
            dispatch(loadNewDMIfNeeded(channel.id));
        } else if (channel && channel.type === Constants.GM_CHANNEL) {
            dispatch(loadNewGMIfNeeded(channel.id));
        }

        const post = threadData.posts[postId];

        if (isCollapsed && isComment(post)) {
            const {data} = await dispatch(focusReplyPost(post, channel, teamId, returnTo, option));
            if (!data) {
                return;
            }
        } else {
            dispatch(focusRootPost(post, channel));
        }

        dispatch(loadChannelsForCurrentUser());
        dispatch(getChannelStats(channelId));
    };
}
