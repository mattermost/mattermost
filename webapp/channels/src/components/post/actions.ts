// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {removePost} from 'mattermost-redux/actions/posts';
import type {ExtendedPost} from 'mattermost-redux/actions/posts';

import {removeDraft} from 'actions/views/drafts';
import {closeRightHandSide} from 'actions/views/rhs';
import {getGlobalItem} from 'selectors/storage';
import {isThreadOpen} from 'selectors/views/threads';

import {StoragePrefixes} from 'utils/constants';

import type {ActionFunc} from 'types/store';

/**
 * This action is called when the deleted post which is shown as 'deleted' in the RHS is then removed from the channel manually.
 * @param post Deleted post
 */
export function removePostCloseRHSDeleteDraft(post: ExtendedPost): ActionFunc<boolean> {
    return (dispatch, getState) => {
        if (isThreadOpen(getState(), post.id)) {
            dispatch(closeRightHandSide());
        }

        const draftKey = `${StoragePrefixes.COMMENT_DRAFT}${post.id}`;
        if (getGlobalItem(getState(), draftKey, null)) {
            dispatch(removeDraft(draftKey, post.channel_id, post.id));
        }

        return dispatch(removePost(post));
    };
}
