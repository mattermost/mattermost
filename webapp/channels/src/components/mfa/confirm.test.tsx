// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {shallow} from 'enzyme';
import React from 'react';

import {redirectUserToDefaultTeam} from 'actions/global_actions';

import Confirm from 'components/mfa/confirm';

import {mountWithIntl} from 'tests/helpers/intl-test-helper';
import Constants from 'utils/constants';

jest.mock('actions/global_actions', () => ({
    redirectUserToDefaultTeam: jest.fn(),
}));

describe('components/mfa/components/Confirm', () => {
    const originalAddEventListener = document.body.addEventListener;

    afterAll(() => {
        document.body.addEventListener = originalAddEventListener;
    });

    test('should match snapshot', () => {
        const wrapper = shallow(<Confirm/>);
        expect(wrapper).toMatchSnapshot();
    });

    test('should submit on form submit', () => {
        const wrapper = mountWithIntl(<Confirm/>);
        wrapper.find('form').simulate('submit');

        expect(redirectUserToDefaultTeam).toHaveBeenCalled();
    });

    test('should submit on enter', () => {
        const map: { [key: string]: any } = {
            keydown: null,
        };
        document.body.addEventListener = jest.fn().mockImplementation((event: string, callback: string) => {
            map[event] = callback;
        });

        mountWithIntl(<Confirm/>);

        const event = {
            preventDefault: jest.fn(),
            key: Constants.KeyCodes.ENTER[0],
        };
        map.keydown(event);

        expect(redirectUserToDefaultTeam).toHaveBeenCalled();
    });
});
