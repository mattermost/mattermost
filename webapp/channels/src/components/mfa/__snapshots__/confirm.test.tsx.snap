// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/mfa/components/Confirm should match snapshot 1`] = `
<div>
  <form
    className="form-group"
    onKeyPress={[Function]}
    onSubmit={[Function]}
  >
    <strong>
      <MemoizedFormattedMessage
        defaultMessage="Set up complete!"
        id="mfa.confirm.complete"
      />
    </strong>
    <p>
      <MemoizedFormattedMessage
        defaultMessage="Your account is now secure. Next time you sign in, you will be asked to enter a code from the Google Authenticator app on your phone."
        id="mfa.confirm.secure"
      />
    </p>
    <button
      className="btn btn-primary"
      type="submit"
    >
      <MemoizedFormattedMessage
        defaultMessage="Okay"
        id="mfa.confirm.okay"
      />
    </button>
  </form>
</div>
`;
